### ISSUE TITLE
rails send_data sending 0 bytes after update to 5.0.26

### ISSUE THREAD

**Issue by fernandoluizao** [phusion/passenger#1763]
Hi,

I have the following code:

```
  def index
     @items = Item.all
     respond_to do |format|
      format.csv do
        csv = CSV.generate({:col_sep => ';'}) do |csv|
          csv << ["Id", "Name"]

          @items.each do |item|
            csv << [
              item.id,
              item.name
            ]
          end
        end
        send_data csv.encode("UTF-16LE"), :filename => 'items.csv'
      end
    end
  end
```

This code was running nice on Passenger 5.0.25 with Apache, CSV is returned correctly.

But after upgrade to 5.0.26, my app is returning a 0 byte file.

I think it may be related to https://github.com/phusion/passenger/commit/99968801d9af84638c2fdd5553d7011b3fe817c0 but I'm not familiar with passenger sources. I'm doing something wrong?

Thanks!

**Comment by FooBarWidget** [comment]
Investigating this now.

**Comment by FooBarWidget** [comment]
I can't reproduce this issue. Can you tell me the raw HTTP response on 5.0.25 and 5.0.26?

**Comment by fernandoluizao** [comment]
No differences between the response of the two versions.

One thing I forgot to mention is that I'm using the deb packages from your ppa.

I can't reproduce this issue using rubygem.

Any idea?

**Comment by FooBarWidget** [comment]
> No differences between the response of the two versions.

I don't understand. Didn't you say that 5.0.26 yields a zero byte response while 5.0.25 yields a correct response?

**Comment by fernandoluizao** [comment]
Sorry about that, here goes the response headers:

5.0.25
```
HTTP/1.1 200 OK
Date: Fri, 04 Mar 2016 10:26:06 GMT
Server: Apache/2.4.7 (Ubuntu)
Cache-Control: private
X-XSS-Protection: 1; mode=block
X-Request-Id: b2acf7c4-db70-446f-a552-575cce106cb4
Content-Disposition: attachment; filename="planos_corte.csv"
Content-Transfer-Encoding: binary
X-Frame-Options: SAMEORIGIN
X-Runtime: 0.148563
X-Content-Type-Options: nosniff
X-Powered-By: Phusion Passenger 5.0.25
Set-Cookie: _mes-slitter_session=OWwvb3N3NkVuRXBEc1V5VFp1a1VlTWU1Slh2NkRNeTA4OXplbUw5ai9UMlpaUXpPSTJ4cXMxb3RmMzhKaG1TbjVYOVFRY3pkdXN6Z2pKZWgwN256b3QwL1YvVllPeHhzbmR2QnBqbW1Bd3cxRVhBU0F4MDNMVEIwNWI3VncrWTRMMjRudDZlQVl0STdHYVpLQzk3S2RNT1ZMSmh0R0FwWEkvQzdRVWs5SENTVzJWbDBsSFVnQWRabkdzN3RCT2lZMGt2UktrcHZmaUs5MFQxQlE5bGNnWk91YjJzdWFjcyt0dGtJMzBGNjZNc3luZG14dU9ZTnJLa0pzZDV3QTNPZy0tazZnUk5XQVAxREx6Ykd2WitscGlQZz09--3270bedf6efd2f1ebc0768290f2f2e35c7c56911; path=/; HttpOnly
ETag: W/"de7f487c0fe25beed308f32a5cb65471"
Status: 200 OK
Keep-Alive: timeout=5, max=98
Connection: Keep-Alive
Transfer-Encoding: chunked
Content-Type: text/csv
```

5.0.26
```
Cache-Control:private
Connection:Keep-Alive
Content-Disposition:attachment; filename="planos_corte.csv"
Content-Length:0
Content-Transfer-Encoding:binary
Content-Type:text/csv
Date:Fri, 04 Mar 2016 10:29:18 GMT
ETag:W/"de7f487c0fe25beed308f32a5cb65471"
Keep-Alive:timeout=5, max=100
Server:Apache/2.4.7 (Ubuntu)
Set-Cookie:_mes-slitter_session=dWFSZi9GWndNa0ZYTGVIRXVBcGhLSVAzTEl6aVRBSHFNQmpQUkp5cElrTFUrajhlbDRvMlNiOUs2UXByMDU0Vm5nTnZPT0grdGhNc1dHU1dRV1pMOGJ2WlY4WVRNS3lSNktBTlhwdjM5b1NaYUlFM3FyVi9sd25PVDdBMW9BVy9SUlVRN2ZiazlxRkZiNHB1Vjh1YlZ0Ykl2NlRmdVR2ZzFmRWhHcTcxTEMyenJEaVE1OUg5Wnd0ODltMU9MMXo4NUZzdzVwYUVKSDFQV2lmS0U4UGZ5NUluNEtGMXJEVUozd2RvZ21rcExFOFZWWmNqNmZrb0JTK29tcXBsS0w5Qy0tbmE2V0xqTENUTFRCTWo0bURVT1h6dz09--5a015a92cc3e431f22d6cd629076d51b16ab976d; path=/; HttpOnly
Status:200 OK
X-Content-Type-Options:nosniff
X-Frame-Options:SAMEORIGIN
X-Powered-By:Phusion Passenger 5.0.26
X-Request-Id:f955e080-210b-4e60-9bb7-79deae62753e
X-Runtime:0.117768
X-XSS-Protection:1; mode=block
```

**Comment by FooBarWidget** [comment]
Can you tell me which Rails version and Ruby version you are on?

We've received a support request from someone else who seems to be experiencing the same issue; he's also encoding to UTF-16LE though he uses 'render :text' instead of 'send_data'. Maybe the issue is related to a specific Rails version or Ruby version.

**Comment by fernandoluizao** [comment]
ruby 2.2.4p230
rails 4.2.5.1

**Comment by fernandoluizao** [comment]
I think this is related to some apache conf... cant reproduce this issue with passenger 5.0.26 standalone (same ruby/rails versions).

**Comment by fernandoluizao** [comment]
I'm closing this issue as it seems that it's not passenger's related.

For now I will stick to v 5.0.25, but I will keep investigating what is causing this, and post updates here.

**Comment by FooBarWidget** [comment]
The other person said that the issue goes way if he downgrades Passenger, and there was a change in 5.0.26 involving the Rack response handler, so it might actually be related. I'm reopening this issue until I've verified that it's not the case.

**Comment by OnixGH** [comment]
It may or may not be related to Apache but do note that the other report we got was also with the that integration mode.

**Comment by OnixGH** [comment]
@fernandoluizao can you check if your log has an error about UTF-16LE encoding when the file is being requested? It should look something like this:

_Exception Encoding::CompatibilityError in Rack response body object (incompatible character encodings: US-ASCII and UTF-16LE)_

If so, could you paste the entire trace?

**Comment by fernandoluizao** [comment]
@OnixGH yes, this error is in the apache logs.

Can I help with some more info or investigation?

**Comment by fernandoluizao** [comment]
Here goes the full stack trace...

```
utils.rb:87 ]: *** Exception Encoding::CompatibilityError in Rack response body object (incompatible character encodings: US-ASCII and UTF-16LE) (process 1577, thread 0x000000013562f8(Worker 1)):
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/ruby_core_io_enhancements.rb:64:in `join'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/ruby_core_io_enhancements.rb:64:in `writev'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/utils/unseekable_socket.rb:127:in `writev'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:298:in `block in process_body'
App 1558 stderr:        from /var/www/mes-slitter/vendor/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/body_proxy.rb:31:in `each'
App 1558 stderr:        from /var/www/mes-slitter/vendor/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/body_proxy.rb:31:in `each'
App 1558 stderr:        from /var/www/mes-slitter/vendor/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/body_proxy.rb:31:in `each'
App 1558 stderr:        from /var/www/mes-slitter/vendor/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/body_proxy.rb:31:in `each'
App 1558 stderr:        from /var/www/mes-slitter/vendor/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/body_proxy.rb:31:in `each'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:295:in `process_body'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:146:in `process_request'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:152:in `accept_and_process_next_request'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:113:in `main_loop'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/request_handler.rb:416:in `block (3 levels) in start_threads'
App 1558 stderr:        from /usr/lib/ruby/vendor_ruby/phusion_passenger/utils.rb:113:in `block in create_thread_and_abort_on_exception'
```

**Comment by OnixGH** [comment]
@fernandoluizao thanks, that looks like what we thought it was. A fix is scheduled: https://github.com/phusion/passenger/commit/c2e32eda9c70ab9a561372de94277efa14c49096

In the meantime you can work around the issue by ensuring that the native util compilation doesn't fail (see the logs for why it's failing, usually it's due to Passenger being started from a user that doesn't have enough access).

**Comment by fernandoluizao** [comment]
@OnixGH that's the compilation error:
```
# current user is: fernando
# mkdir -p /usr/lib/buildout/ruby/ruby-2.2.4-x86_64-linux
Encountered permission error, trying a different directory...
-------------------------------
# mkdir -p /home/fernando/.passenger/native_support/5.0.26/ruby-2.2.4-x86_64-linux
# cd /home/fernando/.passenger/native_support/5.0.26/ruby-2.2.4-x86_64-linux
# /usr/bin/ruby2.2 /usr/lib/src/ruby_native_extension/extconf.rb
/usr/bin/ruby2.2: No such file or directory -- /usr/lib/src/ruby_native_extension/extconf.rb (LoadError)
```
I have installed passenger from your PPA. The packages shouldn't provide all the binaries already compiled? What it's trying to compile?

Many thanks, you guys are the best!

**Comment by OnixGH** [comment]
We provide the essential binaries from our repo. The ruby native extensions that are getting compiled just give a very small performance boost so they are optional.

It just happens that in this case the native code handles the encodings better than the non-native code so having compilation succeed avoids the issue.

5.0.25 had the same code, but maybe something changed in 5.0.26 such that the native compilation stopped working for you, I'll to discuss that with @FooBarWidget. Can you confirm that 5.0.25 did have working compilation for you (with the same environment / user, etc.)?

**Comment by fernandoluizao** [comment]
I confirm that 5.0.25 compiles correctly in the same environment.

**Comment by OnixGH** [comment]
It looks like there's indeed a packaging issue in 5.0.26: `ruby_native_extension` doesn't look like the correct folder, but as a temporary fix to get the compilation going you could just copy the necessary files there anyway, e.g.:
`cp -dpR /usr/share/passenger-enterprise/ruby_extension_source /usr/lib/src/ruby_native_extension`

**Comment by FooBarWidget** [comment]
@fernandoluizao I'm trying to solve the native extension compilation problem. Could you help me by providing some information that I need? Please see this comment: https://github.com/phusion/passenger/issues/1778#issuecomment-198667451

**Comment by FooBarWidget** [comment]
Never mind, I've been able to fix the compilation issue. :)

**Comment by cdcooksey** [comment]
@FooBarWidget What was the resolution to this issue?

**Comment by CamJN** [comment]
This was fixed in 4a0e2812, you can see the message in the changelog for 5.0.27: 

> Fix encoding issue for Ruby apps that resulted in a 0-byte response body. This occurred when the Ruby native support lib was not used and the app outputted an encoding that doesn't mix with UTF-8 (like UTF-16). Closes GH-1763.
